package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,588 @@
+        
+        cdef ns_int i, j
+
+        for i in range(DIM):
+            dx[i] = other[i] - ref[i]
+
+        if self.periodic:
+            for i in range (DIM-1, -1, -1):
+                while dx[i] > self.c_pbcbox.hbox_diag[i]:
+                    for j in range (i, -1, -1):
+                        dx[j] -= self.c_pbcbox.box[i][j]
+
+                while dx[i] <= self.c_pbcbox.mhbox_diag[i]:
+                    for j in range (i, -1, -1):
+                        dx[j] += self.c_pbcbox.box[i][j]
+
+
+    cdef real fast_distance2(self, rvec a, rvec b) nogil:
+        cdef rvec dx
+        self.fast_pbc_dx(a, b, dx)
+        return rvec_norm2(dx)
+
+    
+    cdef real[:, ::1]fast_put_atoms_in_bbox(self, real[:,::1] coords) nogil:
+        cdef ns_int i, m, d, natoms
+        cdef real[:,::1] bbox_coords
+
+        natoms = coords.shape[0]
+        with gil:
+            bbox_coords = coords.copy()
+
+        if self.periodic:
+            if self.is_triclinic:
+                for i in range(natoms):
+                    for m in range(DIM - 1, -1, -1):
+                        while bbox_coords[i, m] < 0:
+                            for d in range(m+1):
+                                bbox_coords[i, d] += self.c_pbcbox.box[m][d]
+                        while bbox_coords[i, m] >= self.c_pbcbox.box[m][m]:
+                            for d in range(m+1):
+                                bbox_coords[i, d] -= self.c_pbcbox.box[m][d]
+            else:
+                for i in range(natoms):
+                    for m in range(DIM):
+                        while bbox_coords[i, m] < 0:
+                            bbox_coords[i, m] += self.c_pbcbox.box[m][m]
+                        while bbox_coords[i, m] >= self.c_pbcbox.box[m][m]:
+                            bbox_coords[i, m] -= self.c_pbcbox.box[m][m]
+        return bbox_coords
+
+#########################
+# Neighbor Search Stuff #
+#########################
+
+cdef class NSResults(object):
+    cdef readonly real cutoff
+    cdef ns_int npairs
+
+    cdef real[:, ::1] coords  # shape: size, DIM
+    cdef real[:, ::1] searchcoords 
+    
+    cdef vector[intvec] indices_buffer
+    cdef vector[realvec] distances_buffer
+    cdef vector[ns_int] pairs_buffer
+    cdef vector[real] pair_distances_buffer
+    cdef vector[real] pair_distances2_buffer
+    
+    def __init__(self, real cutoff, real[:, ::1]coords, real[:, ::1]searchcoords):
+        self.cutoff = cutoff
+        self.coords = coords
+        self.searchcoords = searchcoords
+
+        self.npairs = 0
+     
+    cdef void add_neighbors(self, ns_int beadid_i, ns_int beadid_j, real distance2) nogil:
+        # Important: If this function returns ERROR, it means that memory allocation failed
+
+        self.pairs_buffer.push_back(beadid_i)
+        self.pairs_buffer.push_back(beadid_j)
+        self.pair_distances2_buffer.push_back(distance2)
+        self.npairs += 1
+    
+    def get_pairs(self):

```

Comment:
docs docs docs.  It took me a while to understand what all the different `get_X` mean in this class

Reply:
Yes, I get your point. I was thinking that first we should fix the API and then I would add the docs. However, even for the reviewers to understand, docs are critical. I will add them today.

Reply:
I started to write some docs on my branch... However I am officially on vacation so I do not spend much time of them.
Anyway, I will take some time tomorrow to merge the changes from @ayushsuhane and the very little I have don on the docs.